### PR TITLE
Refactor: introduce state object

### PR DESCRIPTION
**Pull request by betaboon**
continuation of #261 and #263

This PR does two things:
- move the enable- and disable-logic into a dedicated file
- introduce a state-object to replace the singleton-class

This is done to prevent cyclic-imports.

### CODE REVIEW

**Comment by mindflayer** [comment]
As I was previously commenting, I don't particularly like this `Mocket` -> `MocketState` change. It's not a `MocketState` we are dealing with, it's `Mocket` we enable/disable.
I believe it makes things more difficult to write, and expose internals that IMHO should not be exposed.

**Comment by mindflayer** [comment]
OK, now I see there is more. Let me think about that. Thanks!

**Comment by mindflayer** [comment]
While I like the idea of moving the two monkey-patch functions to a different module, I don't think that this `mocket.state.state` is more readable nor understandable than the previous Singleton.

**Comment by betaboon** [comment]
just so i understand, before coming up with another approach:

are you mostly concerned about  about the callsites (ie the `mocket.socket.socket` vs `Mocket.` ?

if so, how about this pattern instead:


```python
from mocket.state import Mocket

def some_function() -> None:
     Mocket.register()
```

by just doing the following in `mocket/state.py`:
```python
Mocket = MocketState()
```

?

**Comment by mindflayer** [comment]
Let's split into different aspects:
- First: what you calling `MocketState` for me is simply `Mocket`, it's not its state;
- Second: what is the reason to convert the Singleton in class/instance when `Mocket` API is designed` to work as a Singleton? If something works like a Singleton, for the sake of readability, isn't it better if it's also implemented as a Singleton?

**Comment by mindflayer** [comment]
> * First: what you calling `MocketState` to me is simply `Mocket`, it's not its state (this is philosophy, I understand you may have a different point of view);

Let me rephrase it: finding a class called `MocketState` is simply confusing, and I don't think it'd be beneficial to the users that already know `Mocket`.

**Comment by betaboon** [comment]
i think I just stepped into the age-old trap of singleton patterns in python.

https://discuss.python.org/t/any-extra-reason-to-not-use-classes-for-singletons/19243/5

where there are two options:
1. have a class bug create a single instance of it
2. have a class that exposes everything through classmethods, that never get's instantiated

you seem to prefer option2, for understandable reasons, while i prefer option 1.
i guess both are valid (aside from you good point against calling it `state`).

I'll try to apply option2 :)

**Comment by mindflayer** [comment]
Honestly: I am not sure I would write it this way today. We did it in 2013 and I don't see any strong reason to change it today.
I can tell you more, I don't normally write Singleton in Python, there is a way better way to write something similar called "Borg".

**Comment by betaboon** [comment]
yeah. i also rarely use singletons. 
but :shrug: the world is full of learning and preferences :)

**Comment by mindflayer** [comment]
The thing here is: we don't need to instantiate Mocket more than once, expecting we always get back the same instance. So, it's not even used as a Singleton, it's just a class implementing some methods, it's pure syntactic sugar.

**Comment by betaboon** [comment]
ok it's back to a singleton-class, with the injection-logic extracted.

i had to change one import `mocket.mocket.MocketSocket` which would lead to cyclic imports.

**Comment by mindflayer** [comment]
> ok it's back to a singleton-class, with the injection-logic extracted.

I really appreciate your patience. :)

**Comment by betaboon** [comment]
of course. you're the maintainer, I'm just a guy.

opensource lives from open, friendly and respectful collaboration :)

**Review by mindflayer — APPROVED** [review]
